ENH: Modified stats.kstest() to use the exact stats.kstwo.sf() where appropriate

pvanmulbregt opened the pull request:
#### Reference issue
gh-9797  Master tracker for K-S Issues. 

#### What does this implement/fix?

This is the third and final in a series of PRs (gh-9719, gh-9753) rewriting
  the Kolmogorov-Smirnov test code.
Add `'exact'`, `'auto'` as values for `mode` in `kstest(...,mode=)`:
  `'exact'` results in calling `stats.kstwo.sf()` instead of the asymptotic `stats.kstwobign.sf()`
For `alternative='two-sided'`,  `'auto'` uses `'exact'` if N is not too large,
  otherwise it uses `'asymp'`.
Change the default value of `mode` from `'asymp'` to `'auto'`.  This results in
  different, but more accurate, values being returned by `stats.kstest()` if the
  `mode` is not specified.
Change `mstats_basic.ks_twosamp()` to call `stats.ks_2samp()` instead of
  performing any calculations itself.
Update returned values in the tutorial and docstrings.

pvanmulbregt commented:
The one Travis CI failure is due to gh-11711.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5297,12 +5329,14 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
-          * 'asymp': use asymptotic distribution of test statistic
+          * 'auto' : uses 'exact' for small size arrays, 'asymp' for large.
+          * 'exact' : uses the exact distribution of test statistic.
+          * 'approx' : approximates the two-sided probability with twice the one-sided probability
```
what is the use case of approx?

pvanmulbregt replied:
`approx` used to return either the asymptotic value `kstwobign.sf(D * np.sqrt(N))`, or twice the one-sided probability `2 * ksone.sf(D, N)`, depending on whether `N > 2666` and other criteria which you asked about below.
Now `approx` always means `2 * ksone.sf(D, N)`.
`auto` will choose one of  [`exact` or `asymp`], never `approx`.
The retention of `approx` is required for backwards compatibility.
I'd expect new  code calling `kstest()` to use the default `mode=auto`.

chrisb83 replied:
OK, maybe add a comment to the docstring to clarify that this option is mainly for backwards compatability and it is not needed in most situations (maybe some users will test how well the approximation works if they are familiar with the test). The approx option is not available in mstats for that reason I assume

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5297,12 +5329,14 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
+    mode : {'auto', 'exact', 'approx', 'asymp'}, optional
         Defines the distribution used for calculating the p-value.
-        The following options are available (default is 'approx'):
+        The following options are available (default is 'auto'):
```
is the mode only relevant for 'two-sided'?

pvanmulbregt replied:
Yes.  The one-sided tests call  `distributions.ksone.sf` and that decides when to switch from exact to an approximation.

chrisb83 replied:
should we mention this in the documentation?

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5378,48 +5412,48 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
+    Dminus = _compute_dminus(cdfvals)
+    D = np.max([Dplus, Dminus])
+    if mode == 'auto':
+        mode = 'exact' if N < 1000000 else 'asymp'
```
1e6 seems to be very large, wouldn't the asymptotic formula work well for much smaller N, e.g. N=1000?

pvanmulbregt replied:
Upon reflection, it's not clear to me whether this switchover needs to exist any more.  The `exact` path calls `kstwo.sf(D, N)` which knows when to approximate with the asymptotic.  The asymptotic likely matches well for large `N`, except for small `D`, those in the range `[0, 2/N]` say, as the exact value approaches `1` faster than the asymptotic.

chrisb83 replied:
agree, one could even consider to use 'exact' as the default option?

chrisb83 replied:
the current version is:
```if mode == 'auto':  # Always select exact
    mode = 'exact'
```
i think auto is not even needed for backwards compatibility since it is introduced in this PR. should we go with exact as the default option or do I miss something?

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5378,48 +5412,48 @@ def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx'):
+        prob = distributions.kstwobign.sf(D * np.sqrt(N))
+    else:
+        # mode == 'approx'
+        prob = 2 * distributions.ksone.sf(D, N)
```
do you know if the old logic including the check `N > 2666 or pval_two > 0.80 - N*0.3/1000` offers any advantage?

pvanmulbregt replied:
That seems to have been a heuristic approximation whose origin is unknown to me (no reference in the code, no mention in any papers I've seen, and IIRC Josef didn't know either.).  Certainly it is less accurate.

chrisb83 replied:
ok, it is better to clean up the implementation in that case

pvanmulbregt commented on scipy/stats/mstats_basic.py:
```diff
@@ -1282,30 +1282,12 @@ def ks_twosamp(data1, data2, alternative="two-sided"):
-
+    alternative = {'t': 'two-sided', 'g':'greater', 'l': 'less'}.get(
+       alternative.lower()[0], alternative)
+    data1 = ma.asarray(data1)
```
Appended the `compressed()` calls onto the `asarray()` calls.

chrisb83 commented on scipy/stats/__init__.py:
```diff
@@ -263,6 +263,7 @@
    brunnermunzel
    combine_pvalues
    jarque_bera
+   Ks_2sampResult
```
for the other tests that return named tulples, we do not mention the output here. do you propose to change that?

pvanmulbregt replied:
By having `mstats_basic` K-S functions call `stats` K-S functions, I no longer needed to expose that named tuple. 
Exposing them all in the API would require taking a look at the names and rationalizing them, making consistent, as that is not the case, from a quick perusal.

pvanmulbregt commented:
The latest push had quite a few changes so I'll put the description here and then add some comments.

--------------
ENH: Modified stats.kstest() to use stats.kstwo.sf().

This is the third and final in a series of PRs (#9719, #9753) rewriting
  the Kolmogorov-Smirnov test code.
Add 'exact', 'auto' as values for mode in stats.kstest(...,mode=):
  'exact' calls stats.kstwo.sf() instead of stats.kstwobign.sf()
For alternative='two-sided',  'auto' uses 'exact' if N is not too large,
  otherwise it uses 'asymp'.
Change the default value of mode from 'asymp' to 'auto'.  This results in
  different, but more accurate, values being returned by stats.kstest() if the
  mode is not specified.
Enhance stats.kstest(data1, data2, ...) so that it performs the 1-sample test if
  given a cdf function, and the 2=sample test if provided two arrays.
Support masked arrays in stats.{kstest,ks_1samp,ks_2samp}.  Previously masked
  arrays were accepted but the wrong answer was returned.
Change mstats_basic.ks_twosamp() to call stats.ks_2samp() instead of
  performing any calculations itself.
Added mstats_basic.ks_2samp() to call stats.ks_2samp().
Added mstats_basic.kstest() to call stats.kstest().
Update returned values in the tutorial and docstrings.
Updated test values in (@pytest.mark.slow) TestKSTwoSamples.testLargeBoth().
Reduced exact-to-asymp switchover from 1000000 to 10000 in stats.kstest().
Added test of stats.kstest()'s multiple code paths in kstest: test_kstest_allpaths().
Added TestCompareWithStats.test_ks_2samp() to check stats/mstats agreement.
Added handling of NaN values in _kstest.py's API routines.
Replaced a manually computed product with np.prod(expr) in stats.ks_2samp().
Added some comments in a stats test function: test_kstest_allpaths().
Moved kstest and ks_2samp next to each other in scipy/stats/stats.py.
Moved _contains_nan from mstats_basic.py to stats.py.
Add tests of the new functionality.

pvanmulbregt commented:
1. The basic premise remains the same: Have K-S testing call the appropriate `stats.distribution.{ksone|kstwo|kstwobign}.cdf` to calculate the p-value of the K-S statistic. 


1. While addressing @chrisb83 's comments I noticed that while `stats.mstats_basic` has some support for KS testing using masked arrays, the basic `stats.stats` implementation of `kstest` also (silently) accepted masked arrays.   It just shunted all the masked values to the end which produced the correct statistic but the the wrong p-value. I added support for masked arrays into  the `stats.stats` KS functions and have the `mstats_basic` functions call the `stats` versions.  This also makes  `mstats_basic` and `stats` kstests agree _a priori_, rather than having to keep them in sync.

1. I also enhanced the signature of `kstest()` to support both one sample and two sample testing.
```
def kstest(data1, data2, args=(), N=20, alternative='two-sided', mode='auto'):
```
`kstest` now decodes the argument possibilities and dispatches to either the one- or two-sample test.

 - `kstest(samples1, samples2,...)` => two-sample testing 
 - `kstest(samples1, cdffunc, args=...)` => one-sample testing, against `cdffunc(*args)`
 - `kstest(samples1, "norm", args=...)` => one-sample testing, against `stats.norm.cdf(*args)` [`"norm"` can be replaced with the name of any distribution in `stats.distributions`.]
 - `kstest(rvsfunc, cdffunc, args=..., N=...)` => one-sample testing, compute a sample of size N using `rvsfunc(*args, size=N)` against `cdffunc(*args)`
 - `kstest("norm", cdffunc, args=..., N=...)` => one-sample testing, compute a sample of size N using `stats.distributions.norm.rvs(*args, size=N)`, against `cdffunc(*args)`. 


1. There are some unusual, all equivalent, additional possibilities which date back to a commit in 2008 3e3e8eaef5ff26148c3f79c925b4d20b15e05988.

 - `kstest("norm", "norm", args=..., N=...)` => one-sample testing, compute a sample of size N using `distributions.norm.rvs(*args, size=N)`, against `distributions.norm.cdf(*args)`
 - `kstest("norm", "", args=..., N=...)` => same
 - `kstest("norm", False, args=..., N=...)` => same
 - `kstest("norm", [], args=..., N=...)` => same
 - `kstest("norm", 0, args=..., N=...)` => same
 -  In fact, if `data1` is a string then any value of `data2` that evaluates False has this result.

What is the purpose of this variation, telling `kstest` to generate a random sample from a distribution and then comparing the sample against that same distribution's CDF?  Is it  just for some ease of testing or is there a real-world use case? @josef-pkt perhaps you have some context here?

pvanmulbregt commented:
Closes gh-11884 as a side-effect.

pvanmulbregt commented:
The `kstest, ks_2samp, ks_1samp` family of functions work for both masked arrays and non-masked arrays. Found a way to import the functions from `stats` into the `mstats` namespace, needed for backward compatibility. (`mstats_basic` and `stats` are unfortunately circularly dependent.)

pvanmulbregt commented:
Spoke too soon.  Importing `ks_2samp` from `stats` into `mstats` requires that any references in the docstring are also visible or the doc builds fail. will restore the shims in `mstats`.

chrisb83 commented:
@pvanmulbregt let me know if I should continue my review. (unfortunately the diff of `stats.py` is displayed in a way that is very hard to read, some functions are shown as deleted and reappear somewhere else.)

pvanmulbregt commented:
>  let me know if I should continue my review. (unfortunately the diff of `stats.py` is displayed in a way that is very hard to read, some functions are shown as deleted and reappear somewhere else.)

@chrisb83 Yes, it is fine to continue.  (The strange diff in stats may be because I moved the `kstest` function below the worker functions, and it contains the docstrings.)

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    >>> stats.kstest(x,'norm', alternative = 'greater')
+    (0.0072115233216311081, 0.98531158590396395)
+
+    Don't reject equal distribution against alternative hypothesis: greater
```
I did not understand this comment. 

Why is asymp used below?

pvanmulbregt replied:
Removed `asymp` in the example below.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+
+    Don't reject equal distribution against alternative hypothesis: greater
+
+    >>> stats.kstest(x,'norm', mode='asymp')
```
there are a few missing spaces below

pvanmulbregt replied:
Done.

chrisb83 replied:
thanks, when I did my first review, I did not see that these PEP8 issues already were in the existing documentation

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    >>> stats.kstest(x,'norm', mode='asymp')
+    (0.12464329735846891, 0.08944488871182088)
+
+    *Testing t distributed random variables against normal distribution*
```
the example below is much easier to follow than the one above that has less comments on the idea behind the tests

pvanmulbregt replied:
I believe that the docstring for this function was copied verbatim from the docstriong for the existing `stats.kstest`. If you have replacement doc examples you prefer now I' be happy to use them! If not, I'd like to make those changes in a different PR, as it will likely affect multiple functions.

chrisb83 commented:
one more observation: in mstats, `ks_twosamp` is be a candidate for deprecation given that there is already `ks_2samp`?

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -6183,71 +6284,197 @@ def ks_2samp(data1, data2, alternative='two-sided', mode='auto'):
+    # (name of distribution, empty string, ...)
+    # (name of distribution, False, ...)
+
+    # Returns xvals, yvals, cdf
```
which combinations are possible for xvals?

pvanmulbregt replied:
xvals is array_like.  Updated documentation.

chrisb83 commented:
> kstest now decodes the argument possibilities and dispatches to either the one- or two-sample test. 
kstest(samples1, samples2,...) => two-sample testing
kstest(samples1, cdffunc, args=...) => one-sample testing, against cdffunc(*args)
kstest(samples1, "norm", args=...) => one-sample testing, against stats.norm.cdf(*args)

The above use case are the normal ones in my view. The other combinations make the API very complicated and hard to follow and from the current documentation, I am not sure if the behaviour of things like `kstest("norm", False, args=..., N=...)` are clear.

In any case, since this PR just builds on the exising API, so I am fine with allowing such cases.

pvanmulbregt commented on scipy/stats/_ksstats.py:
```diff
@@ -414,25 +414,20 @@ def _kolmogn(n, x, cdf=True):
+        if nxsquared >= 2.2:
+            prob = 2 * scipy.special.smirnov(n, x)
+            return _clip_prob(prob)
+        # Fall through and compute 1-cdf
```
Should be "Compute SF as 1-CDF."

pvanmulbregt commented on scipy/stats/_ksstats.py:
```diff
@@ -414,25 +414,20 @@ def _kolmogn(n, x, cdf=True):
+        cdfprob = _kolmogn_DMTW(n, x, cdf=True)
+    else:
+        cdfprob = _kolmogn_PelzGood(n, x, cdf=True)
+    return _select_and_clip_prob(cdfprob, 1.0 - cdfprob, cdf=cdf)
```
Not quite.  If `n > 140` and `n*x*x < 2.2`, the SF  is computed as 1-CDF.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    Parameters
+    ----------
+    x : array_like
+        a 1-D array of observations of random variables.
```
Done.

pvanmulbregt commented:
> > kstest now decodes the argument possibilities and dispatches to either the one- or two-sample test.
> > kstest(samples1, samples2,...) => two-sample testing
> > kstest(samples1, cdffunc, args=...) => one-sample testing, against cdffunc(*args)
> > kstest(samples1, "norm", args=...) => one-sample testing, against stats.norm.cdf(*args)
> 
> The above use case are the normal ones in my view. The other combinations make the API very complicated and hard to follow and from the current documentation, I am not sure if the behaviour of things like `kstest("norm", False, args=..., N=...)` are clear.
> 
> In any case, since this PR just builds on the exising API, so I am fine with allowing such cases.

Personally I'd be in favor of dropping support for these cases.  Nobody jumped to explain the use cases after https://github.com/scipy/scipy/pull/11710#issuecomment-616149862 and they do make the API complicated and difficult to maintain.

pvanmulbregt commented:
> one more observation: in mstats, `ks_twosamp` is a candidate for deprecation given that there is already `ks_2samp`?

`ks_twosamp` is in the API.  But it would make sense to deprecate the `XX_twosamp` routines. E.g. `ks_twosamp` in `stats` and `mstats`, `ttest_onesamp` in `mstats`.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    Returns
+    -------
+    statistic : float
+        KS test statistic, either D, D+ or D-.
```
Done.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+    ----------
+    x : array_like
+        a 1-D array of observations of random variables.
+    cdf : callable
```
Right now `ks_1samp` doesn't accept a string.  `kstest(x, ...)` does, and it converts the string to a cdf function before passing down to `ks_1samp`.  Since `ks_1samp` is not in the API currently, it doesn't need to support all the myriad combinations of `kstest`.
If one wants the convenience of strings, use the wrapper `kstest` which figures out everything.

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -5819,7 +5689,185 @@ def chisquare(f_obs, f_exp=None, ddof=0, axis=0):
+
+    *Test against one-sided alternative hypothesis*
+
+    Shift distribution to larger values, so that ``cdf_dgp(x) < norm.cdf(x)``:
```
Weird.  `cdf_dgp` has been there for at least 8 years and I have no idea what it means.  Removed it and replaced with `CDF(x) < norm.cdf(x)`

chrisb83 commented:
> Personally I'd be in favor of dropping support for these cases. Nobody jumped to explain the use cases after #11710 (comment) and they do make the API complicated and difficult to maintain.

I agree.

chrisb83 commented on scipy/stats/tests/test_mstats_basic.py:
```diff
@@ -1436,6 +1436,60 @@ def test_obrientransform(self):
+                        res3 = stats.ks_1samp(xm, stats.norm.cdf, alternative=alternative, mode=mode)
+                        assert_equal(np.asarray(res1), np.asarray(res3))
+
+    def test_kstest_1samp(self):
```
is this the same test as in line 1439 (copy-paste error?)

tylerjereddy commented:
@chrisb83 @pvanmulbregt Just a reminder we'd like to branch 1.5.x in about 6 days--looks like there's still plenty of activity here so maybe it will converge on time?

pvanmulbregt commented:
* Added comments to the K-S test functions to make clearer their purpose.

* Dropped support for the mostly undocumented cases which served to only make testing `kstest` use slightly fewer characters but had little real-world applicability and made maintenance more difficult. These include:

```
kstest("norm", "", args=..., N=...)
kstest("norm", False, args=..., N=...)
kstest("norm", [], args=..., N=...)
kstest("norm", 0, args=..., N=...)
```

pvanmulbregt commented:
@chrisb83 Updated based upon your latest comments. Did I miss any?

chrisb83 commented:
@pvanmulbregt yes, thanks, my comments above have been addressed. 

I took a look at the testing and I think there should be two basic test cases for `ks_1samp`:

- check names of attributes (copy&paste from TestKSTwoSamples)
- one test of the exact option and the alternatives, just to be sure that this is correctly implemented. I generated values for comparison in R below (R version 3.4.1 (2017-06-30))

```
x <- c(-1.23, 0.06, -0.60, 0.17, 0.66, -0.17, -0.08, 0.27, -0.98, -0.99)

res <- ks.test(x, "pnorm", exact = TRUE)
res['statistic'] # 0.2935801
res['p.value'] # 0.2934085

res <- ks.test(x, "pnorm", alternative = 'greater', exact = TRUE)
res['statistic'] # 0.2935801 (D+)
res['p.value'] # 0.1469888

res <- ks.test(x, "pnorm", alternative = 'less', exact = TRUE)
res['statistic'] # 0.1093486 (D-)
res['p.value'] # 0.7327689
```

I don't think more testing is required since there is a lot of testing of  `stats.ksone`. Let me know if you think my suggestion makes sense.

Apart from this point, this PR is ready to be merged in my view.

One follow-up PR could enhance the documentation (e.g. examples). The null hypothesis of `kstest` is incorrectly described as 

> Under the null hypothesis, the two distributions are identical, F(x)=G(x). The alternative hypothesis can be either ‘two-sided’ (default), ‘less’ or ‘greater’. 

However, the null varies depending on alternative. These points are not related to this PR.

pvanmulbregt commented:
The single CI failure is due to `tools/lint_diff.py`. However the errors and lines mentioned don't seem to match up with the actual code.

```
$ tools/lint_diff.py
./scipy/stats/tests/test_stats.py:2908:27: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2908:29: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2911:28: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2911:31: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2938:27: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2938:29: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2941:28: E231 missing whitespace after ','
./scipy/stats/tests/test_stats.py:2941:31: E231 missing whitespace after ','
8       E231 missing whitespace after ','
```
E.g. test_stats.py:2908-2911 looks fine 
https://github.com/scipy/scipy/blob/e663f2abff06033020e6d069a67cf67d2339793e/scipy/stats/tests/test_stats.py#L2908-L2911

@person142  This branch was created before` lint_diff.py` existed.  Does it have any difficulties in that situation?

pvanmulbregt commented:
@chrisb83  I added the comparison with R's values from above and took a look at the testing code. 
I suspect that some of the testing is redundant,  now that `stats` and `mstats` have the same underlying implementation. 
The testing of `kstest` can be simplified to test just what it does in terms of wrapping up `ks_1samp` and `ks_2samp`, and leave the heavy lifting to the testing of `ks_1samp` and `ks_2samp`. 
As you point out, the testing of `scipy.stats.stats.ks_1samp` and `scipy.stats.ksone` likely have overlap, or could be partitioned differently.

I think the current status is:
* The functionality is complete, the implementation is final.
* The test suite covers what needs to be tested.
* The testing of `kstest`, `ks_1samp`, `ksone` may test more than is needed and could be rationalized.
* The existing documentation could use some review.

My preference would be to place any changes to the testing/documentation in a new PR after 1.5, as that will take some thinking and review and won't happen before the 1.5 fork.  
And to merge this PR in to make it available for 1.5 (once the `lint_diff.py` issue is understood.)

chrisb83 commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2875,41 +2876,112 @@ def test_friedmanchisquare():
+
+    def test_agree_with_r(self):
+        # comparing with some values from R
+        x = np.linspace(-1,1,9)
```
do the missing spaces after , cause the whitespace error? there are a few more cases above and below

chrisb83 commented:
> My preference would be to place any changes to the testing/documentation in a new PR after 1.5, as that will take some thinking and review and won't happen before the 1.5 fork.

yes, let's get this PR merged before the 1.5 fork. My comments above were addressed, I will just do one final iteration this weekend.

chrisb83 commented on scipy/stats/mstats_basic.py:
```diff
@@ -1282,34 +1300,35 @@ def ks_twosamp(data1, data2, alternative="two-sided"):
 
 
-ks_2samp = ks_twosamp
+def kstest(data1, data2, args=(), alternative='two-sided', mode='auto'):
```
in ks_2samp / ks_twosamp, data1 and data2 are replaced by x and y whereas kstest still uses data1 and data2. is that intendend?

chrisb83 replied:
the change in ks_twosamp could even be a problem with backwards compatability?

pvanmulbregt replied:
Good spot.  The arguments could be provided by name to `ks_2samp`, even though they are required args.  In `kstest` it is not an issue as `kstest` was not exposed before now.

chrisb83 commented on scipy/stats/stats.py:
```diff
@@ -6056,7 +6147,7 @@ def ks_2samp(data1, data2, alternative='two-sided', mode='auto'):
     Parameters
     ----------
-    data1, data2 : sequence of 1-D ndarrays
+    x, y : sequence of 1-D ndarrays
```
same question as above: can we rename these parameters?

pvanmulbregt replied:
I'll change them back in `stats.ks_2samp`.
But that raises an interesting question on `kstest`.
The signature of `kstest` changed:
```
def kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='approx')`
-->
def kstest(data1, data2, args=(), N=20, alternative='two-sided', mode='auto')
```

The default `mode` should not be an issue.  But the name change of `rvs, cdf` -> `data1, data2` might be.  The changed names reflect the enhanced nature of the function.  It supports the old use cases of `kstest`, which handled one sample testing.  It also supports two sample testing via `kstest(rvs1, rvs2)`, in which case `rvs2` is definitely not a cdf of any kind.

@rgommers @tylerjereddy Is it permitted to change the names of positional arguments of a public API function?
I did a quick Google search for "kstest(rvs=..." and found a couple of examples specifying `rvs` by keyword, so my assumption is that changing the argument names at this time is not the thing to do.  Just document very clearly that `cdf` can be more than just a `cdf`. Is that correct?

chrisb83 commented:
Two more observation:

- the parameters data1 / x and data2 / y are used in the code and if I am not mistaken, the renaming could create an issue with backwards compatability
- the code now returns a named tuple KStestresult for all variations of the test. can we do such a change in view of the existing functionality that is in use?

chrisb83 commented:
Last but not least: let's not forget to link the issues collected in #9797 that will be closed by this PR

WarrenWeckesser commented:
>  Is it permitted to change the names of positional arguments of a public API function?

No, that's a backwards incompatible API change.

pvanmulbregt commented:
Restored original names for backwards compatibility.

pvanmulbregt commented:
Closes gh-8748 gh-8341 gh-9797

tylerjereddy commented:
CI all green & core dev approval, merging.

tylerjereddy commented:
I've manually closed & milestoned the cross-linked issues above

tylerjereddy commented:
Maybe suggest a release note entry directly in the WIP PR: https://github.com/scipy/scipy/pull/12221

pvanmulbregt commented:
I updated the Wiki with an entry for `kstest`.
